_posts/2019-04-10-imperatywnie-vs-funkcyjnie.md

```diff
@@ -0,0 +1,118 @@
+---
+layout:    post
+title:     "Programowanie imperatywne vs funkcyjne"
+date:      2019-04-10 9:00:00 +0100
+published: false
+author:    Łukasz Sroczyński
+tags:
+    - backend
+    - kotlin
+    
+---
+
+Programowanie funkcyjne w ostatnich latach stało się bardzo popularne. Umożliwia nam to pisanie kodu, który jest czystszy, bezpieczniejszy, a przede wszystkim łatwo testowalny. Oczywiście FP ma również wady z czego jedną z nich może być zbyt duże pójście w abstrakcję co powoduje, że kod staje się niezrozumiały. Omówimy sobie w takim razie jakie są różnice pomiędzy imperatywnym podejściem, a tym funkcyjnym. 
+
+## Różnica między kodem imperatywnym, a funkcyjnym?
+Imperatywnie znaczy tyle, że mutujemy zmienne, sprawdzamy warunki (if, else) oraz robimy wiele potencjalnie niebezpiecznych operacji. Na sam koniec tworzymy wynik. W takim świecie ciągle zmieniającego się stanu bardzo łatwo o błąd. Można sobie wyobrazić, że taki kod jest dosłownie szyty na miarę. Program składa się wtedy z rzeczy, które "coś robią". Ma on jakiś stan wejściowy, przejściowy oraz wyjściowy. W funkcyjnym podejściu metoda przyjmuje jakieś generyczne argumenty i zwraca wynik. Jest to możliwe dzięki kompozycji funkcji. Tak też programy funkcjne składają się z części, które "są czymś".
```

Comment:
No nie do końca... programowanie imperatywne nie polega na tym, że "mutujemy zmienne", "sprawdzamy warunki" czy "robimy potencjalnie niebezpieczne operacje" tylko na tym, że przechodzimy za pomocą ciągu instrukcji z jednego stanu systemu w inny.

Warto w ogóle zaznaczyć, że programowanie imperatywne wydaje się naturalnym ze względu na to, że komputery działają w sposób imperatywny. Pewnie  warto o tym wspomnień gdyż, mimo że napiszesz swój program w "bezpieczny" funkcyjny sposób, to podczas jego wykonania i tak będzie działać w "niebezpieczny" imperatywny sposób - na przykład będzie przypisywać coś do rejestrów czy korzystać ze stosu

Reply:
Warto też od razu wspomnieć o efektach ubocznych. Brakuje mi stwierdzenia, że wywołanie funkcji można zastąpić jej wartością dla danych argumentów. Może warto też wspomnieć nieco o funkcji w sensie matematycznym, bo tak naprawdę programowanie funkcyjne opiera się o tą strasznie teoretyczna matematyczna definicję ;)

Reply:
Piszesz o programowaniu imperatywnym w sposób, który sugeruje, że to zło wcielone. Postaraj się o tym napisać w sposób neutralny

Reply:
Chciałem trochę zmienić, a napiałem od nowa (wyszło trochę więcej tekstu).
Zerknij jeszcze raz, czy jest sens dopisywania czegoś jeszcze.